src/uu/factor/src/numeric.rs

```diff
@@ -172,55 +193,152 @@ impl Arithmetic for Montgomery {
     }
 }
 
+// NOTE: Trait can be removed once num-traits adds a similar one;
+//       see https://github.com/rust-num/num-traits/issues/168
+pub(crate) trait OverflowingAdd: Sized {
+    fn overflowing_add_(self, n: Self) -> (Self, bool);
+}
+macro_rules! overflowing {
+    ($x:ty) => {
+        impl OverflowingAdd for $x {
+            fn overflowing_add_(self, n: Self) -> (Self, bool) {
+                self.overflowing_add(n)
+            }
+        }
+    };
+}
+overflowing!(u32);
+overflowing!(u64);
+overflowing!(u128);
+
+pub(crate) trait Int:
+    Display + Debug + PrimInt + OverflowingAdd + WrappingNeg + WrappingSub + WrappingMul
+{
+    fn as_u64(&self) -> u64;
+    fn from_u64(n: u64) -> Self;
+    #[cfg(debug_assertions)]
+    fn as_u128(&self) -> u128;
+    #[cfg(debug_assertions)]
+    fn from_u128(n: u64) -> Self;
+}
+
+pub(crate) trait DoubleInt: Int {
+    type Double: Int;
+
+    fn as_double(self) -> Self::Double;
+    fn from_double(n: Self::Double) -> Self;
+}
```

Comment:
Maybe something like ...

```suggestion
pub(crate) trait VariablePrecisionInt: Int {
    type HighPrecision: Int;

    fn as_high_precision(self) -> Self::HighPrecision;
    fn from_high_precision(n: Self::HighPrecision) -> Self;
}
```

Reply:
@rivy Yes, `Double` is an integer type with double the width of `Self`.
`HighPrecision` (or `DoublePrecision` for that matter) isn't quite correct, as those are integers; what is increased isn't the precision, but the range (which is squared).

However, this was rather helpful, I think I have an idea for better naming now.